Update K8up Documentation (#205)
vshn/k8up: docs/modules/ROOT/pages/tutorials/tutorial.adoc

= Backing up Kubernetes Clusters with K8up

> One of the most common questions we got from companies moving to Kubernetes has always had to do with backups: how can we ensure that the information in our pods and services can be quickly and safely restored in case of problems?

Author: Yes, makes sense. We also try to write one sentence per line as per https://syn.tools/syn/references/style-guide.html#_asciidoc_style

$ cd /var/lib/mysql # Attention! Different path!
$ mv data /
$ rm -r *
$ cp -R /data/mariadb-pvc/* .
....

[start=5]
. Delete the MariaDB pod, and a new one will be immediately created according to the deployment: `kubectl delete pod mariadb-56cd59d499-jrkr7`

Now refresh your WordPress browser window and you should see the previous state of the WordPress installation restored, working and looking as expected!

.WordPress website restored
image::tutorial/wordpress-restored.png[]

[[step_6]]
=== Cleaning up the backup pods

NOTE: The operations of this step can be executed at once using the `scripts/6_delete_restore_pods.sh` script.

Whenever K8up performs a backup, it creates a job pod; the one we created previously can be manually deleted using the command:

`kubectl delete -f k8up/restore-wordpress.yaml`

[[step_7]]
=== Scheduling regular backups

NOTE: The operations of this step can be executed at once using the `scripts/7_schedule.sh` script.

Instead of performing backups manually, you can also set a schedule for backups. This requires specifying the schedule in `cron` format.

[source,yaml]
....
backup:
  schedule: '*/2 * * * *'    # backup every 2 minutes
  keepJobs: 4

>   promURL: http://minio:9000

Author: That's a hack to workaround issues. Will be fixed in the rewrite later.

....

TIP: Use https://crontab.guru/[crontab.guru] to help you set up complex schedule formats in `cron` syntax.

The schedule can also specify `archive` and `check` tasks to be executed regularly.

[source,yaml]
....
archive:
  schedule: '0 0 1 * *'       # archive every week
  restoreMethod:
    s3:
      endpoint: http://minio:9000
      bucket: archive
      accessKeyIDSecretRef:
        name: backup-credentials
        key: username
      secretAccessKeySecretRef:
        name: backup-credentials
        key: password
check:
  schedule: '0 1 * * 1'      # monthly check
  promURL: http://minio:9000
....

Run the `kubectl apply -f k8up/schedule.yaml` command. This will setup an automatic schedule to backup the PVCs every 5 minutes (for minutes that are divisors of 5).

Wait for at most 2 minutes, and run the `restic snapshots` to see more backups piling up in the repository.

> TIP: Running the `watch restic snapshots` command will give you a live console with your current snapshots on a terminal window, updated every 2 seconds.

Reviewer: ```suggestion
TIP: Running the `watch restic snapshots` command will rerun restic every 2 seconds and update the output so that you don't have to do it yourself.
```

[[step_8]]
=== Cleaning up the cluster

NOTE: The operations of this step can be executed at once using the `scripts/8_stop.sh` script.

> When you are done with this tutorial, just execute the `minikube stop` command to shut the cluster down. You can also `minikube delete` it, if you would like to get rid of it completely.

Reviewer: ```suggestion
When you are done with this tutorial, just execute the `minikube stop` command to shut the cluster down. If you would like to get rid of it completely, run `minikube delete`.
```

== Conclusion

> We hope that this walkthrough has given you a good overview of K8up and its capabilities. But it can do much more than that! We haven't talked about the archive, prune, and check commands, or about the backup of any data piped to `stdout` (called "Application Aware" backups.) You can check these features in the https://k8up.io/[K8up documentation website] where they are described in detail.

Reviewer: ```suggestion
We hope that this walkthrough provided you a quick overview of K8up and its capabilities. But K8up can do more than that! We have only briefly mentioned the archive, prune, and check commands and haven't talked about the backup of any data piped to `stdout` (called "Application Aware" backups.) You can check these features on the https://k8up.io/[K8up documentation website] where they are described in detail.
```
